File: spoj/1437-PT07Z/solution_1437.cpp
#include <bits/stdc++.h>
using namespace std;

int longestPathDFS(int arr[][2], int nodes, int root = -1)
{

    bool fromMax = root != -1;

    int stack[nodes], top = -1;

    bool visitedNodes[nodes];
    memset(visitedNodes, false, nodes);

    stack[++top] = arr[0][0];

    if (fromMax)
        stack[top] = root;

    visitedNodes[stack[top] - 1] = true;
    int visited = 1;
    int max = -1, maxPos = -1;

    while (visited < nodes && top != -1)
    {
        int i = 0;
        while (i < nodes - 1)
        {
            int first = arr[i][0];
            int second = arr[i][1];

            if (first == stack[top] && !visitedNodes[second - 1])
            {
                visited++;
                stack[++top] = second;
                visitedNodes[second - 1] = true;
                i = 0;
            }
            else if (second == stack[top] && !visitedNodes[first - 1])
            {
                visited++;
                stack[++top] = first;
                visitedNodes[first - 1] = true;
                i = 0;
            }
            else
                i++;
        }
        if (maxPos < top)
        {
            maxPos = top;
            max = stack[top];
        }

        --top;
    }

    return fromMax ? maxPos : max;
}

int main()
{
    int nodes;
    cin >> nodes;

    int arr[nodes - 1][2];

    for (int i = 0; i < nodes - 1; i++)
    {
        for (int j = 0; j < 2; j++)
        {
            cin >> arr[i][j];
        }
    }

    int maxVal = longestPathDFS(arr, nodes);
    int maxPath = longestPathDFS(arr, nodes, maxVal);
    cout << maxPath;

    return 0;
}

File: spoj/1437-PT07Z/solution_1437.py
# time limit exception
def longest_path_dfs(edges, nodes, root=-1):
    fromRoot = root != -1

    visitedNodes = [False for x in range(nodes + 1)]
    stack = [edges[0][0]]

    if fromRoot:
        stack[0] = root

    visitedNodes[stack[0]]
    visited = 1
    maxPos = -1
    maxVal = -1

    while visited < nodes and len(stack) != 0:
        i = 0
        while i < nodes - 1:
            first = edges[i][0]
            second = edges[i][1]
            if first == stack[-1] and not visitedNodes[second]:
                visitedNodes[second] = True
                i = 0
                visited += 1
                stack.append(second)
            elif second == stack[-1] and not visitedNodes[first]:
                visitedNodes[first] = True
                i = 0
                visited += 1
                stack.append(first)
            else:
                i += 1
        if maxPos < (len(stack) - 1):
            maxPos = len(stack) - 1
            maxVal = stack[-1]
        stack.pop()

    return maxPos if fromRoot else longest_path_dfs(edges, nodes, maxVal)


if __name__ == "__main__":
    nodes = int(input())
    edges = []

    for x in range(nodes - 1):
        edges.append(list(map(int, input().split(" "))))
    print(longest_path_dfs(edges, nodes))
# 5
# 1 2
# 1 3
# 2 4
# 2 5
